File: reference/src/front-end/blocks.ts
import type { MarkdownNode, MarkdownRoot } from "../markdown/parseMarkdown.js";
import type { SourceSpan, TangleImport, TangleParam } from "../model.js";
import { spanFromNode } from "./sourceMap.js";

export function collectLinks(file: string, root: MarkdownRoot): TangleImport[] {
  const imports: TangleImport[] = [];

  function walk(node: MarkdownNode): void {
    if (node.type === "link" && node.url) {
      const alias = plainText(node).trim();
      if (alias && node.url.endsWith(".md")) {
        imports.push({ alias, target: node.url, span: spanFromNode(file, node) });
      }
    }

    for (const child of node.children ?? []) {
      walk(child);
    }
  }

  walk(root);
  return imports;
}

export function parseParamItem(text: string, span: SourceSpan): TangleParam {
  // Phase 6c: 修正正则以支持 `name`: (Type) 形式（无 description，仅有类型注解）。
  // 原正则要求 `\s+` 前置空格，但 `: (Config)` 中 `:\s*` 已消耗空格，导致
  // `(Config)` 被当作 description。改为 `\s*` 并允许 description 为空，与
  // Rust parse_param_item 行为一致（Rust 用 rfind('(')/rfind(')') 提取最后
  // 括号对作为 typeName，不要求 description 非空）。
  const match = text.match(/^`([^`]+)`:\s*(.*?)(?:\s*\(([^)]+)\))?$/);
  if (!match || !match[1]) {
    throw new Error(`Invalid Tangle parameter item: ${text}`);
  }

  const result: TangleParam = {
    name: match[1],
    description: (match[2] ?? "").trim(),
    span
  };
  if (match[3]) {
    result.typeName = match[3];
  }
  return result;
}

export function isTangleCodeBlock(node: Pick<MarkdownNode, "type" | "lang">): boolean {
  return node.type === "code" && node.lang === "@tangle";
}

export function plainText(node: MarkdownNode): string {
  if (typeof node.value === "string") {
    return node.value;
  }

  return (node.children ?? []).map(plainText).join("");
}

File: compiler/tangle-cli/src/frontend/blocks.rs
use crate::markdown::MarkdownNode;
use crate::model::{SourceSpan, TangleImport, TangleParam};

pub fn collect_links(file: &str, nodes: &[MarkdownNode]) -> Vec<TangleImport> {
    let mut imports = vec![];
    collect_links_recursive(file, nodes, &mut imports);
    imports
}

fn collect_links_recursive(file: &str, nodes: &[MarkdownNode], out: &mut Vec<TangleImport>) {
    for node in nodes {
        if node.node_type == "link" {
            if let Some(ref url) = node.url {
                if url.ends_with(".md") {
                    let alias = node.children.iter()
                        .find(|c| c.node_type == "text")
                        .and_then(|c| c.value.clone())
                        .unwrap_or_else(|| "unknown".to_string());
                    if let Some(span) = node.to_span(file) {
                        out.push(TangleImport { alias, target: url.clone(), span });
                    }
                }
            }
        }
        collect_links_recursive(file, &node.children, out);
    }
}

pub fn parse_param_item(text: &str, span: &SourceSpan) -> Option<TangleParam> {
    let text = text.trim();
    if !text.starts_with('`') { return None; }
    let end_tick = text[1..].find('`')?;
    let name = &text[1..=end_tick];
    let after_name = text[end_tick + 2..].trim();
    let after_name = after_name.strip_prefix(':')?.trim();
    let (description, type_name) = if let Some(open) = after_name.rfind('(') {
        if let Some(close) = after_name.rfind(')') {
            if close > open {
                (after_name[..open].trim().to_string(), Some(after_name[open + 1..close].to_string()))
            } else { (after_name.to_string(), None) }
        } else { (after_name.to_string(), None) }
    } else { (after_name.to_string(), None) };
    Some(TangleParam { name: name.to_string(), description, type_name, span: span.clone() })
}

pub fn is_tangle_code_block(node: &MarkdownNode) -> bool {
    node.node_type == "code" && node.lang.as_deref() == Some("@tangle")
}

pub fn plain_text(node: &MarkdownNode) -> String {
    if node.node_type == "text" { return node.value.clone().unwrap_or_default(); }
    node.children.iter().map(plain_text).collect::<Vec<_>>().join("")
}
